daft/expressions/expressions.py

```diff
@@ -3554,3 +3563,101 @@ class ExpressionEmbeddingNamespace(ExpressionNamespace):
     def cosine_distance(self, other: Expression) -> Expression:
         """Compute the cosine distance between two embeddings."""
         return Expression._from_pyexpr(native.cosine_distance(self._expr, other._expr))
+
+
+class ExpressionBinaryNamespace(ExpressionNamespace):
+    def length(self) -> Expression:
+        """Retrieves the length for a binary string column.
+
+        Example:
+            >>> import daft
+            >>> df = daft.from_pydict({"x": [b"foo", b"bar", b"baz"]})
+            >>> df = df.select(df["x"].binary.length())
+            >>> df.show()
+            ╭────────╮
+            │ x      │
+            │ ---    │
+            │ UInt64 │
+            ╞════════╡
+            │ 3      │
+            ├╌╌╌╌╌╌╌╌┤
+            │ 3      │
+            ├╌╌╌╌╌╌╌╌┤
+            │ 3      │
+            ╰────────╯
+            <BLANKLINE>
+            (Showing first 3 of 3 rows)
+
+        Returns:
+            Expression: an UInt64 expression with the length of each binary string in bytes
+        """
+        return Expression._from_pyexpr(native.binary_length(self._expr))
+
+    def concat(self, other: Expression | bytes) -> Expression:
+        r"""Concatenates two binary strings.
+
+        Example:
+            >>> import daft
+            >>> df = daft.from_pydict(
+            ...     {"a": [b"Hello", b"\\xff\\xfe", b"", b"World"], "b": [b" World", b"\\x00", b"empty", b"!"]}
+            ... )
+            >>> df = df.select(df["a"].binary.concat(df["b"]))
+            >>> df.show()
+            ╭────────────────────╮
+            │ a                  │
+            │ ---                │
+            │ Binary             │
+            ╞════════════════════╡
+            │ b"Hello World"     │
+            ├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+            │ b"\\xff\\xfe\\x00" │
+            ├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+            │ b"empty"           │
+            ├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
+            │ b"World!"          │
+            ╰────────────────────╯
+            <BLANKLINE>
+            (Showing first 4 of 4 rows)
+
+        Args:
+            other: The binary string to concatenate with, can be either an Expression or a bytes literal
+
+        Returns:
+            Expression: A binary expression containing the concatenated strings
+        """
+        if isinstance(other, bytes):
+            other = lit(other)
+        return Expression._from_pyexpr(native.binary_concat(self._expr, other._expr))
+
+    def substr(self, start: Expression | int, length: Expression | int | None = None) -> Expression:
```

Comment:
Since binary values do not technically have to be substrings, maybe we should rename this to slice. PyArrow calls it the same: https://arrow.apache.org/docs/python/generated/pyarrow.compute.binary_slice.html